### PR TITLE
Question: What is the correct interpretation of LaneBitmask?

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
While enabling sub-reg liveness for AArch64, we ran into a register allocator issue, which leads us to ask the question: What is the correct interpertation of the bits in LaneBitmask ?

(1) The documentation of LaneBitmask suggests that if A == B that A overlaps with B, but suggests nothing that guarantees that A fully overlaps with B.
(2) The code pointed out in this PR makes the assumption that if A == B, that A fully overlaps with B.

AArch64 defines 16-bit H0 as being the larger register of 8-bit B0, e.g.
```
  def B0    : AArch64Reg<0,   "b0">, DwarfRegNum<[64]>;
  ... 

  let SubRegIndices = [bsub] in {
    def H0    : AArch64Reg<0,   "h0", [B0]>, DwarfRegAlias<B0>;
    ..
  }
```
It doesn't define the top bits of H0 as an artificial register (e.g. B0_HI).

In the reproducer the register allocator has to insert a spill. Because 'bsub' (8-bit) and 'hsub' (16-bit), 'ssub' (32-bit), etc. all have the same bitmask, it picks a register class for the COPY instruction using the first subreg-idx that matches. This results in a copy of a 32-bit value, rather than the (desired) full 128-bit value.

If H0 *would* be defined as B0 and B0_HI (and similar for all other registers), then the following condition evaluates true only for the largest subreg index:

  // Early exit if we found a perfect match.
  if (SubRegMask == LaneMask) {
    BestIdx = Idx;
    break;
  }

There are several ways to fix this issue;
* If statement (2) is correct, then we could define the top bits with an artificial B0_HI register. It would also be nice for TableGen to emit some extra information that LLVM can assert, so that enabling subreg liveness doesn't result in spurious errors.
* If statement (1) is correct however, the code (as pointed out in this PR) simply contains a bug that we should fix.

Can someone with more knowledge of the register allocator provide guidance? Thank you!

### CODE REVIEW

**Comment by arsenm** [comment]
As it is today, I believe the intent is statement 1. However, I think we would be in a better place if we did require 2. We do have a number of issues whenever there are pieces of registers that aren't fully covered by a defined register, and it is harder to reason about.

We probably need to require fully covering artificial register definitions to fully move the infrastructure to using register units (e.g. replacing livein tracking with regunits instead of registers, and for using regunit masks instead of regmasks)

**Comment by jayfoad** [comment]
> As it is today, I believe the intent is statement 1. However, I think we would be in a better place if we did require 2.

Agreed.

> We do have a number of issues whenever there are pieces of registers that aren't fully covered by a defined register, and it is harder to reason about.
> 
> We probably need to require fully covering artificial register definitions to fully move the infrastructure to using register units (e.g. replacing livein tracking with regunits instead of registers, and for using regunit masks instead of regmasks)

This was discussed in some detail on #95926 (also see #96146). Also also see https://github.com/llvm/llvm-project/pull/79831#issuecomment-1943746260.

**Comment by sdesmalen-arm** [comment]
Thanks for the quick feedback!

> We probably need to require fully covering artificial register definitions to fully move the infrastructure to using register units (e.g. replacing livein tracking with regunits instead of registers, and for using regunit masks instead of regmasks)

Does the approach of using register units (instead of registers) have any impact on the existence or future use of the LaneBitmask?

> As it is today, I believe the intent is statement 1. However, I think we would be in a better place if we did require 2. We do have a number of issues whenever there are pieces of registers that aren't fully covered by a defined register, and it is harder to reason about.

I have some local patches that define registers/subreg-indices for the top bits, but then ran into an issue that the `uint64_t` to represent the `LaneBitmask` is no longer sufficient. It can be changed to use an `APInt` but I'm not sure if there's a way to define the BitWidth on a per-target basis (at least, not without passing some Subtarget to the LaneBitmask constructor). I haven't yet tried to see what the impact is on compile-time/memory-usage, or whether there is a way to limit this on a per-target basis, but this is a downside of having to define the top bits.

**Comment by arsenm** [comment]
> Does the approach of using register units (instead of registers) have any impact on the existence or future use of the LaneBitmask?
> 

No. In an ideal world you could directly map bits 


> 
> I have some local patches that define registers/subreg-indices for the top bits, but then ran into an issue that the `uint64_t` to represent the `LaneBitmask` is no longer sufficient. 


How big are these registers? AMDGPU is sitting right at the 64-bit limit now

**Comment by jayfoad** [comment]
> I have some local patches that define registers/subreg-indices for the top bits, but then ran into an issue that the uint64_t to represent the LaneBitmask is no longer sufficient.

Is this for the regular AArch64 integer registers? My understanding is:

> AArch64 registers work like this: b10 aliases the low 8 bits of h10 which aliases the low 16 bits of s10 which aliases the low 32 bits of d10 which aliases the low 64 bits of q10 which aliases the low 128 bits of z10.

So I would expect you would need subregisters for:
- bits 7..0
- bits 15..8
- bits 31..16
- bits 63..32
- bits 127..64
- bits 255..128

I.e. roughly log2(bitwidth) of them.

**Comment by sdesmalen-arm** [comment]
> > I have some local patches that define registers/subreg-indices for the top bits, but then ran into an issue that the uint64_t to represent the LaneBitmask is no longer sufficient.
> 
> Is this for the regular AArch64 integer registers? My understanding is:
> 
> > AArch64 registers work like this: b10 aliases the low 8 bits of h10 which aliases the low 16 bits of s10 which aliases the low 32 bits of d10 which aliases the low 64 bits of q10 which aliases the low 128 bits of z10.
> 
> So I would expect you would need subregisters for:
> 
> * bits 7..0
> * bits 15..8
> * bits 31..16
> * bits 63..32
> * bits 127..64
> * bits 255..128
> 
> I.e. roughly log2(bitwidth) of them.

That's right, however it is the register tuples that make things take up space, as it replicates the bits for each subvector in the tuple. It ends up with something like this:
```
qqsub_0: bsub, bsub_hi, hsub, hsub_hi, ... qsub, qsub_hi
qqsub_1: bsub, bsub_hi, hsub, hsub_hi, ... qsub, qsub_hi
```
and similar for other tuples, where each sub register must be separately addressable and cannot reuse the bits (otherwise qqsub_0:bsub and qqsub_1:bsub would alias for example). Then there are also bits to represent the indices in GPR registers a set of bits needed to express the (sub)tiles in a matrix register. With all of these together in the same bitmask, we run out of the 64 bits.

File: llvm/lib/CodeGen/TargetRegisterInfo.cpp [hunk 1]
```diff
@@ -521,26 +521,62 @@ bool TargetRegisterInfo::getCoveringSubRegIndexes(
...
+    // However, in the code here the assumption is made that if
+    // 'SubRegMask == LaneMask', then the subregister must overlap *fully*.
+    // This is not the case for the way AArch64 registers are defined
+    // at the moment.
+    //
+    // Which interpretation is correct?
+    //
+    // A) If the meaning is 'if the bits are equal, the sub-registers overlap but
+    //    not necessarily fully', then we should fix the code in this function (in
+    //    a better way than just disabling it).
```
Comment (qcolombet) [inline llvm/lib/CodeGen/TargetRegisterInfo.cpp:557]: Like @arsenm said this is the correct interpretation.

Comment (sdesmalen-arm) [reply llvm/lib/CodeGen/TargetRegisterInfo.cpp:557]: Thanks for confirming! Now that this is cleared up, I'll close this PR.

**Pull request closed without merging**